File: lab2/bag.py
'''
Created on Mar 11, 2019

@author: Andrea
'''

class Bag:

    # creates a new, empty Bag
    def __init__(self):
        self.elements = []
        self.frequency = []

    # adds a new element to the Bag
    '''
    Complexity: O(n);  n = nr elem unice
    '''
    def add(self, e):
        if e in self.elements:
            for i in range(0, len(self.elements)):
                if(e == self.elements[i]):
                    self.frequency[i] += 1
                    return
        else:
            self.elements.append(e);
            self.frequency.append(1);

    # removes one occurrence of an element from a Bag
    # returns True if an element was actually removed (the Bag contained the element e), or False if nothing was removed
    """
    Complexity: O(n)
    """
    def remove(self, e):
        if e in self.elements:
            for i in range(0, len(self.elements)):
                if e == self.elements[i]:
                    if (self.frequency[i] != 1):
                        self.frequency[i] -= 1
                        return True
                    else:
                        self.elements.pop(i)
                        self.frequency.pop(i)
                        return True
        else:
            return False

    # searches for an element e in the Bag
    # returns True if the Bag contains the element, False otherwise
    """
    Complexity: O(n)
    """
    def search(self, e):
        if e in self.elements:
            for i in range(0, len(self.elements)):
                if e == self.elements[i]:
                    if self.frequency[i] > 0:
                        return True
        return False

    # counts and returns the number of times the element e appears in the bag
    '''
    Complexity: O(n)
    '''
    def nrOccurrences(self, e):
        if e in self.elements:
            for i in range(0, len(self.elements)):
                if e == self.elements[i]:
                    return self.frequency[i]
        return 0
    
    # returns the size of the Bag (the number of elements)
    '''
    Complexity: Theta(n)
    '''
    def size(self):
        count = 0;
        for i in range(0, len(self.elements)):
            count += self.frequency[i];
        return count

    # returns True if the Bag is empty, False otherwise
    """
    Complexity: theta(n)
    """
    def isEmpty(self):
        if self.size() == 0:
            return True
        return False

    # returns a BagIterator for the Bag
    """
    Complexity Thetha(1)
    """
    def iterator(self):
        return BagIterator(self)


class BagIterator:

    #creates an iterator for the Bag b, set to the first element of the bag, or invalid if the Bag is empty
    def __init__(self, b):
        self.__bag = b
        self.__current = 0
        self.__freq = 0
        
        
    # returns True if the iterator is valid
    def valid(self):
        if self.__current + 1 < len(self.__bag.elements):
            return True
    
    
    # returns the current element from the iterator.
    # throws ValueError if the iterator is not valid
    def getCurrent(self):
        return self.__bag.elements[self.__current]
    
    # moves the iterator to the next element
    #throws ValueError if the iterator is not valid
    def next(self):
        if self.__current == len(self.__bag.elements):
            raise ValueError
        else:
            if self.__freq == 1:
                self.__current += 1
                self.__freq = self.__bag.elements[self.__current]
            else:
                self.__freq -= 1
    
    # sets the iterator to the first element from the Bag
    def first(self):
        self.__current = 0
        self.__freq = 0

File: lab1/Bag.cpp
#include "Bag.h"
#include "BagIterator.h"
#include<math.h>

//complexity: theta(1)
Bag::Bag()
{
}

//complexity: theta(1)
Bag::~Bag()
{

}


//complexity: o(n) , n - numar elemente unice
void Bag::add(TElem e)
{
	TElem i;
	TElem pos = -1;
	for (i = 0; i < this->unique.getSize(); i++)
	{
		if (e == this->unique[i])
		{
			pos = i;
			break;
		}
	}
	if (pos != -1)
	{
		this->positions.add(pos);
	}
	else
	{
		this->unique.add(e);
		this->positions.add(this->unique.getSize() - 1);
	}
}


//complexity: o(n) , n -numar elemente din bag
bool Bag::remove(TElem e)
{
	TElem i;
	int freq = 0;
	int pos = -1;
	int pos_unique = -1;
	for (i = 0; i < this->positions.getSize(); i++)
	{
		if (e == this->unique[this->positions[i]])
		{
			/*
			1 2 2 2 2 5 7 2 2

			u: 1 2 5 7
			p: 1 2 2 2 2 3 4 2 2 
			*/
			freq++;
			pos = i;
			pos_unique = this->positions[i];
		}
	}

	if (freq > 1)
	{
		this->positions[pos] = this->positions[this->positions.getSize() - 1];
		this->positions.removeLast();
		return true;
	}

	if (freq == 1)
	{
		int old_pos = this->unique.getSize() - 1;
		this->unique[pos_unique] = this->unique[old_pos];
		this->positions[pos] = this->positions[this->positions.getSize() - 1];
		this->positions.removeLast();
		for (i = 0; i < this->positions.getSize(); i++)
		{
			if (this->positions[i] == old_pos)
				this->positions[i] = pos_unique;
		}
		this->unique.removeLast();

		return true;
	}
	return false;
}

//complexity: o(n) , n - nr elem unice 
bool Bag::search(TElem e)
{
	TElem i;
	for (i = 0; i < this->unique.getSize(); i++)
	{
		if ( this->unique[i] == e)
		{
			return 1;
		}
	}
	return 0;
}


void Bag::intersection(const Bag& b)
{
	int i = 0, nOcurrent, nObag, diff;
	for (i = 0; i < this->unique.getSize(); i++)
	{
		nOcurrent = this->nrOccurrences(this->unique[i]);
		nObag = b.nrOccurrences(this->unique[i]);
		diff =nOcurrent - nObag;
		if (diff > 0)
		{
			int j;
			for (j = 0; j < diff; j++)
				this->remove(this->unique[i]);
		}
	}
}

//complexity: theta(1)
int Bag::size() const
{
	return this->positions.getSize();
}

//complexity: theta(n) , n - nr elemente bag
int Bag::nrOccurrences(TElem e) const
{
	TElem i;
	int freq = 0;
	for (i = 0; i < this->positions.getSize(); i++)
	{
		int position = this->positions[i];
		if (e == this->unique[position])
		{
			freq++;
		}
	}
	return freq;
}

//complexity: theta(1)
bool Bag::isEmpty() const
{
	if (this->positions.getSize() == 0)
		return true;
	return false;
		
}

//complexity: theta(1) 
BagIterator Bag::iterator() const
{
	BagIterator it( *this );
	return it;
}


